### PR TITLE
fix: concurrent event triggering unnecessary flush call

### PR DESCRIPTION
**Pull request by MoumitaM**
## Description of the change

> Concurrent events are triggering unnecessary flush calls resulting in a Bad request error. 
So, moved the logic to remove batch items from the queue before waiting for the pending flush to avoid the above scenario.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,8 +2,8 @@
   "name": "@rudderstack/rudder-sdk-node",
   "version": "2.0.3",
   "description": "Rudder Node SDK",
-  "main": "./cjs/index.js",
-  "module": "./esm/index.js",
+  "main": "dist/cjs/index.js",
+  "module": "dist/esm/index.js",
```
Comment (bardisg) [inline package.json:6]: Why this was changed? This should be the path inside the dist folder once the package.json is in the package

**Review by bardisg — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (MoumitaM) [reply package.json]: Adressed

File: src/index.js [hunk 2]
```diff
@@ -611,6 +611,10 @@ class Analytics {
       return Promise.resolve();
     }
 
+    // remove items from the queue before awaiting for the pending flush
+    // to avoid concurrent events triggering unnecessary flush API calls
+    const items = this.queue.splice(0, this.flushAt);
+
```
Comment (saikumarrs) [inline src/index.js:617]: We should avoid this. If the pending flush throws an error, we'll lose these events from the queue.
Instead, restore Ln 623 and check for the length of `items` before proceeding further. Make sure to invoke the callback when it's empty.

**Review by saikumarrs — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 2 above)
Comment (MoumitaM) [reply src/index.js]: Adressed

File: src/index.js [hunk 3]
```diff
@@ -621,6 +621,11 @@ class Analytics {
     }
 
     const items = this.queue.splice(0, this.flushAt);
+    // Do not proceed in case the items array is empty
```
Comment (bardisg) [inline src/index.js:624]: if this fix resolves the issue we need it also in service worker for v1.1 and v3

**Review by bardisg — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

**Pull request merged**